clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1201,11 +1250,71 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
   }
 };
 
+// A type to check the validity of passing union with accessor/sampler/stream
+// member as a kernel argument types.
+class SyclKernelUnionBodyChecker : public SyclKernelFieldHandler {
+  static constexpr const bool VisitUnionBody = true;
```

Comment:
I think your problem is that SyclKernelFieldHandler needs this line as well, except =false instead.

I think that results in the union-filtering not actually doing anything.

Reply:
>> think that results in the union-filtering not actually doing anything.
Yes, Thanks Erich.

Reply:
Done

Reply:
I can't find where it is done, that might be the reason why everything is decomposed.

Reply:
Yep, it should be in the base class, not just the other checker.  WHich is why the tests are still wrong.

Reply:
Done.